src/app/components/shared/input/date/date.component.ts

```diff
@@ -0,0 +1,310 @@
+        (change)="onInput($event)"
+        (dateSelect)="onSelect($event)"
+        (blur)="onTouched()"
+      />
+      <button
+        type="button"
+        class="btn btn-outline-secondary"
+        (click)="dateInput.toggle()"
+      >
+        <fa-icon [icon]="['fas', 'calendar']"></fa-icon>
+      </button>
+      <div *ngIf="showError" class="invalid-feedback">
+        {{ errors[0] }}
+      </div>
+    </div>
+  `,
+})
+export class DateComponent
+  implements ControlValueAccessor, Validator, OnChanges
+{
+  /** Label next to input */
+  @Input() public label: string;
+  /** Minimum date allowed in selection */
+  @Input() public minDate: Date;
+  /** Maximum date allowed in selection */
+  @Input() public maxDate: Date;
+  /** Is input required */
+  @Input() public required = false;
+  /** Is input disabled */
+  @Input() public disabled = false;
+
+  // TODO Add range selection which sets the min and max dates
+
+  /** Current value as an NgbDate structure */
+  public date: NgbDate;
+  /** Has value been set */
+  public dirty: boolean;
+  /** Current value */
+  public value: Date;
+  /** Has input been touched */
+  public touched: boolean;
+
+  /** Errors for current input */
+  public _errors: ValidationErrors;
+  /** Potential error messages */
+  public errorTypes: Map<string, () => string>;
+  /** Length of input yyyy-mm-dd */
+  public inputLength = 10;
+  /** Maximum allowed date */
+  public maxDateStruct: NgbDateStruct;
+  /** Minimum allowed date */
+  public minDateStruct: NgbDateStruct = { year: 1970, month: 1, day: 1 };
+
+  public constructor(
+    private formatter: NgbDateParserFormatter,
+    private calendar: NgbCalendar
+  ) {
+    this.maxDateStruct = this.calendar.getToday();
+    this.errorTypes = new Map([
+      ["required", () => `${this.label} is required`],
+      [
+        "minLength",
+        () => "The time should follow the format yyyy-mm-dd, e.g. 2022-12-01",
+      ],
+      [
+        "delimiter",
+        () =>
+          "Input must include a '-' between the date segments, e.g 2022-12-01",
+      ],
+      [
+        "hour",
+        () =>
+          `Year must be between ${this.minDateStruct.year} and ${this.maxDateStruct.year}`,
+      ],
+      ["month", () => "Month must be between 1 and 12"],
+      ["day", () => "Day must be between 1 and 31"],
+      [
+        "minDate",
+        () => `Date must be after ${this.formatter.format(this.minDateStruct)}`,
+      ],
+      [
+        "maxDate",
+        () =>
+          `Date must be before ${this.formatter.format(this.maxDateStruct)}`,
+      ],
+      ["invalid", () => "Invalid date selected"],
+    ]);
+  }
+
+  public ngOnChanges(): void {
+    const convertDateToStruct = (date: Date) => ({
+      year: date.getFullYear(),
+      month: date.getMonth() + 1,
+      day: date.getDate(),
+    });
+    const convertStructToDate = (struct: NgbDateStruct) =>
+      new Date(struct.year, struct.month - 1, struct.day);
```

Comment:
neither of these methods have anything to with structs. fix the naming.

Reply:
The name of the type converting to and from is `NgbDateStruct`, i could change to `convertDateToNgbDateStruct` if you would prefer?

Reply:
I'd avoid the term struct altogether. 

While there is [a proposal](https://github.com/tc39/proposal-structs) for JS to get structs, it currently does not have them. Just because a library uses a bad name doesn't mean we're going to get sucked in as well.

Also they have a converter: https://ng-bootstrap.github.io/#/components/datepicker/api#NgbDateAdapter